### PR TITLE
Enable backtrace with MSVC

### PR DESCRIPTION
**Pull request by peterjc123**
Make it possible to report the C++ exceptions in console.

### CODE REVIEW

File: c10/util/Backtrace.cpp [hunk 1]
```diff
@@ -194,7 +214,87 @@ std::string get_backtrace(
...
+      get_module_name(module, back_trace[i_frame]);
+      stream << module << " @ <unknown line number>";
+    }
+    stream << "]" << std::endl;
+
+    delete line;
+    line = nullptr;
+  }
+
+  delete[] back_trace;
```
Comment (peterjc123) [inline c10/util/Backtrace.cpp:294]: Use unique_ptr

File: c10/util/Backtrace.cpp [hunk 2]
```diff
@@ -194,7 +214,87 @@ std::string get_backtrace(
...
+    stream << " [";
+    if (with_line) {
+      stream << line->FileName << " @ " << line->LineNumber;
+    } else {
+      get_module_name(module, back_trace[i_frame]);
+      stream << module << " @ <unknown line number>";
+    }
+    stream << "]" << std::endl;
+
+    delete line;
```
Comment (peterjc123) [inline c10/util/Backtrace.cpp:290]: Use unique_ptr

**Comment by ezyang** [comment]
Just a note: by default we suppress output of C++ backtraces; I'm wondering if on Windows we should flip the default so we provide them by default.

File: c10/util/Backtrace.cpp [hunk 3]
```diff
@@ -115,7 +120,22 @@ c10::optional<FrameInformation> parse_frame_information(
   frame.function_name = demangle(mangled_function_name.c_str());
   return frame;
 }
-
+} // anonymous namespace
+#elif defined(_MSC_VER)
+namespace {
+const int max_name_len = 256;
+void get_module_name(char* module, void* addr) {
```
Comment (ezyang) [inline c10/util/Backtrace.cpp:127]: nit: this is kind of frustratingly low level for C++. Wonder if we can just return a `std::string` here.

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
This got unlanded because it's not Werror clean on internal builds:

```

stderr: xplat\caffe2\c10\util\Backtrace.cpp(279,9): error: address of array 'module' will always evaluate to 'true' [-Werror,-Wpointer-bool-conversion]
    if (module) {
    ~~  ^~~~~~
```

**Comment by ezyang** [comment]
Also, apparently, this doubled the time Windows test take, apparently. CC @yns88 who might be able tos ay more.

**Comment by peterjc123** [comment]
To make it faster, maybe we could do `SymInitialize` at startup and `SymCleanup` before quit only once. Could you please tell me how to do that, @ezyang?

**Comment by ezyang** [comment]
I think the easiest way to do it is probably some sort of Meyer's singleton. So we'll have

```
struct SymbolRAII {
  SymbolRAII() {
  ... call SymInitialize ...
  }
  ~SymbolRAII() {
    ... call SymCleanup ...
  }
}

void initialize_symbols() {
  static SymbolRAII x;
}
```

**Pull request closed without merging**